flatsurf/geometry/gl2r_orbit_closure.py

```diff
@@ -0,0 +1,940 @@
+                seen.add(h2)
+                seen.add(h3)
+        return faces
+
+    def __repr__(self):
+        return "GL(2,R)-orbit closure of dimension at least %d in %s (ambient dimension %d)" % (self._U_rank, self.ambient_stratum(), self.d)
+
+    def holonomy(self, v):
+        r"""
+        Return the holonomy of ``v`` (with respect to the basis)
+        """
+        return self.V(v) * self.H
+
+    def holonomy_dual(self, v):
+        return self.V(v) * self.Hdual
+
+    def tangent_space_basis(self):
+        return self._U[:self._U_rank].rows()
+
+    def lift(self, v):
+        r"""
+        Given a vector in the "spanning set basis" return a vector on the full basis of
+        edges.
+
+        EXAMPLES::
+
+            sage: from flatsurf import polygons, translation_surfaces, similarity_surfaces
+            sage: from flatsurf import GL2ROrbitClosure  # optional: pyflatsurf
+
+            sage: S = translation_surfaces.mcmullen_genus2_prototype(4,2,1,1,0)
+            sage: O = GL2ROrbitClosure(S)  # optional: pyflatsurf
+            sage: u0,u1 = O.tangent_space_basis()  # optional: pyflatsurf
+            sage: v0 = O.lift(u0)  # optional: pyflatsurf
+            sage: v1 = O.lift(u1)  # optional: pyflatsurf
+            sage: span([v0, v1])  # optional: pyflatsurf
+            Vector space of degree 9 and dimension 2 over Number Field in l with defining polynomial x^2 - x - 8 with l = 3.372281323269015?
+            Basis matrix:
+            [                         1                          0                         -1   (1/4*l-1/4 ~ 0.59307033) (-1/4*l+1/4 ~ -0.59307033)                          0 (-1/4*l+1/4 ~ -0.59307033)                          0 (-1/4*l+1/4 ~ -0.59307033)]
+            [                         0                          1                         -1    (1/8*l+7/8 ~ 1.2965352) (-1/8*l+1/8 ~ -0.29653517)                         -1 (3/8*l-11/8 ~ -0.11039450) (-1/2*l+3/2 ~ -0.18614066) (-1/8*l+1/8 ~ -0.29653517)]
+
+        This can be used to deform the surface::
+
+            sage: T = polygons.triangle(3,4,13)
+            sage: S = similarity_surfaces.billiard(T)
+            sage: S = S.minimal_cover("translation").erase_marked_points()
+            sage: O = GL2ROrbitClosure(S)  # optional: pyflatsurf
+            sage: for d in O.decompositions(4, 20):  # optional: pyflatsurf
+            ....:     O.update_tangent_space_from_flow_decomposition(d)
+            ....:     if O.dimension() == 4:
+            ....:         break
+            sage: d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf
+            sage: dreal = d1/132 + d2/227 + d3/280 - d4/201  # optional: pyflatsurf
+            sage: dimag = d1/141 - d2/233 + d4/230 + d4/250  # optional: pyflatsurf
+            sage: d = [O.V2((x,y)).vector for x,y in zip(dreal,dimag)]  # optional: pyflatsurf
+
+        TODO: This is waiting for https://github.com/flatsurf/flatsurf/issues/145::
```

Comment:
This might work now…

Reply:
The example above this comment is now failing
```
from flatsurf import polygons, translation_surfaces, similarity_surfaces
from flatsurf import GL2ROrbitClosure
T = polygons.triangle(3,4,13)
S = similarity_surfaces.billiard(T)
S = S.minimal_cover("translation").erase_marked_points()
O = GL2ROrbitClosure(S)  # optional: pyflatsurf
for d in O.decompositions(4, 20):  # optional: pyflatsurf
    O.update_tangent_space_from_flow_decomposition(d)
    if O.dimension() == 4:
        break
d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf
```
gives
```
Traceback (most recent call last)
<ipython-input-2-603dc29281d6> in <module>()
      7     if O.dimension() == Integer(4):
      8         break
----> 9 d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf

<ipython-input-2-603dc29281d6> in <listcomp>(.0)
      7     if O.dimension() == Integer(4):
      8         break
----> 9 d1,d2,d3,d4 = [O.lift(b) for b in O.tangent_space_basis()]  # optional: pyflatsurf

/home/vincent/programming/flatsurf/sage-flatsurf/flatsurf/geometry/gl2r_orbit_closure.py in lift(self, v)
    621         u = vector(self.V2.base_ring(), n + 1)
    622         u[:k] = v
--> 623         return A.solve_right(u)
    624 
    625     def absolute_homology(self):

/opt/sage/local/lib/python3.7/site-packages/sage/matrix/matrix2.pyx in sage.matrix.matrix2.Matrix.solve_right (build/cythonized/sage/matrix/matrix2.c:8345)()
    831 
    832         if not self.is_square():
--> 833             X = self._solve_right_general(C, check=check)
    834         else:
    835             try:

/opt/sage/local/lib/python3.7/site-packages/sage/matrix/matrix2.pyx in sage.matrix.matrix2.Matrix._solve_right_general (build/cythonized/sage/matrix/matrix2.c:9685)()
    954             # Have to check that we actually solved the equation.
    955             if self*X != B:
--> 956                 raise ValueError("matrix equation has no solutions")
    957         return X
    958 

ValueError: matrix equation has no solutions
```

Reply:
The problem to have this example works is that the `+` operation works in `pyflatsurf` surfaces and that `GL2ROrbitClosure` does not accept them. Is that something we want to allow? In order to allow that, one would need
- to be able to check whether the input `surface` in `GL2ROrbitClosure.__init__` is a `FlatTriangulation` (which is a template in cppyy, not something accepted by `isinstance`)
- extract the base ring from the `surface` in order to build the various vector spaces necessary for the `GL2ROrbitClosure` class

Reply:
The alternative is of course to be able to go back and forth between `pyflatsurf` and `sage-flatsurf` in a more transparent fashion.

Reply:
See also 4014171